### PR TITLE
Feature/remove eager loading

### PR DESCRIPTION
**Pull request by coorasse**
Using code from @elabs and continuing on what I wrote here: https://github.com/CanCanCommunity/cancancan/pull/171#issuecomment-129334100 that is my suggestion.
Note that I fixed it only for AR4

### CODE REVIEW

File: lib/cancan/model_adapters/active_record_adapter.rb [hunk 1]
```diff
@@ -65,6 +65,7 @@ def database_records
           @model_class.where(nil).merge(override_scope)
         elsif @model_class.respond_to?(:where) && @model_class.respond_to?(:joins)
           if mergeable_conditions?
+            puts conditions
```
Comment (bmulholland) [inline lib/cancan/model_adapters/active_record_adapter.rb]: You'll want to remove the outputs in this file :)

Comment (coorasse) [inline lib/cancan/model_adapters/active_record_adapter.rb]: :+1:

**Comment by Senjai** [comment]
Could we write some more tests to verify the correct sql is being generated? This is a more complex PR and may have some edge cases I may not be aware of. Going to pull it down and review it more thoroughly.

**Comment by coorasse** [comment]
Would you prefer to reduce the amount of code and depend directly from this gem? https://github.com/elabs/ar_outer_joins

Is an old gem with just 15 stars so I preferred to include the code directly but we can eventually remove it and add the dependency. what do you think?

**Comment by mrloop** [comment]
**Configurable includes** In some scenarios I'd like to be able to specify per model what gets included to reduce further sql requests I know I'm going to make when looking at a resources related models.

**Comment by mrloop** [comment]
Ah look somebody has already suggested this #228

**Comment by coorasse** [comment]
You can specify what to include already with `includes(:table_a, :table_b)` directive. Which is even more clear.

**Comment by mrloop** [comment]
@coorasse but I'm loading the resource via `load_and_authorize_resource`. How do I tell cancan to `.includes(:table_a, :table_b)` during `load_and_authorize_resource` find?

**Comment by mrloop** [comment]
Ok, your just getting a scope from load_and_authorize_resource, thanks @coorasse

**Comment by coorasse** [comment]
Hello @Senjai. I think this PR brings an important improvement and, at the same time, change in cancancan.
I can rely on this gem https://github.com/elabs/ar_outer_joins in order to avoid writing more tests here.

I already use the following changes in different projects and I had no problems, even in complex ones (https://github.com/coorasse/Airesis/tree/master/app/cancan).

I really think using `includes` to perform outer_joins is a bad choice and cancancan is taking the responsability of preloading stuff when is not required --> performance problems in case of deep relations.
Let the user preload their stuff and let cancancan just check for permissions.

**Comment by derosm2** [comment]
+1

**Comment by lowjoel** [comment]
@coorasse https://github.com/nerde/left_join seems a little newer. Do you want to use that instead?

**Comment by Senjai** [comment]
@coorasse Want to get on a hangout at some point for this one?

**Comment by coorasse** [comment]
Hi @Senjai, get in touch with me via hangout at coorasse@gmail.com if you want

**Comment by kaspernj** [comment]
+1

**Comment by kaspernj** [comment]
Once the following PR is merged into Rails, it should be possible to just do a `.left_outer_joins` instead in Rails 5: https://github.com/rails/rails/pull/30410

**Comment by aliou** [comment]
Any news on this feature ?

**Comment by coorasse** [comment]
I think this will fall in the next Major version (3.0) where I plan to deprecate also AR4. Therefore this PR will be simplified by a lot and rely on left_outer_join by AR5

File: lib/cancan/model_adapters/ar_outer_joins/ar_outer_joins.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+require "cancan/model_adapters/ar_outer_joins/join_builder"
+require "cancan/model_adapters/ar_outer_joins/join"
+
+module ArOuterJoins
+  def outer_joins(*joins)
```
Comment (kaspernj) [inline lib/cancan/model_adapters/ar_outer_joins/ar_outer_joins.rb]: In Rails 5 we got `.left_joins`. The duplicate table in this regard was fixed in https://github.com/rails/rails/pull/30410.

Maybe rename this to `left_joins` and only use if it Rails version is below Rails 5.2?

**Comment by bmulholland** [comment]
What's the status of v3.0?

**Comment by coorasse** [comment]
Closed in favor of #508
